doc/design/libraries.md

```diff
@@ -0,0 +1,90 @@
+| libbitcoin_common      | Home for common functionality shared by different executables and libraries. Similar to `libbitcoin_util`, but higher-level (see [Dependencies](#dependencies). |
+| libbitcoin_consensus   | Stable, backwards-compatible consensus functionality used by `libbitcoin_node` and `libbitcoin_wallet` and also exposed as a [shared library](../doc/shared-libraries.md). |
+| libbitcoinconsensus    | Shared library build of static `libbitcoin_consensus` library |
+| libbitcoin_kernel      | Consensus "engine" and support library used for validation by `libbitcoin_node` and also exposed as a [shared library](../doc/shared-libraries.md). |
+| libbitcoinqt           | GUI functionality used by `bitcoin-qt` and `bitcoin-gui` executables |
+| libbitcoin_ipc         | IPC functionality used by `bitcoin-node`, `bitcoin-wallet`, `bitcoin-gui` executables to communicate when [`--enable-multiprocess`](multiprocess.md) is used. |
+| libbitcoin_node        | P2P and RPC server functionality used by `bitcoind` and `bitcoin-qt` executables. |
+| libbitcoin_util        | Home for common functionality shared by different executables and libraries. Similar to `libbitcoin_common`, but lower-level (see [Dependencies](#dependencies). |
+| libbitcoin_wallet      | Wallet functionality used by `bitcoind` and `bitcoin-wallet` executables. |
+| libbitcoin_wallet_tool | Lower-level wallet functionality used by `bitcoin-wallet` executable. |
+| libbitcoin_zmq         | [ZeroMQ](../zmq.md) functionality used by `bitcoind` and `bitcoin-qt` executables. |
+
+## Conventions
+
+- Most libraries are internal libraries and have APIs which are completely unstable! There are few or no restrictions on backwards compatibility or rules about external dependencies. Exceptions are `libbitcoin_consensus` and `libbitcoin_kernel` which have external interfaces documented at [../shared-libraries.md](../shared-libraries.md).
+
+- Generally speaking each library should correspond to a source directory and a namespace. Source code organization is a work in progress, so it is true that some namespaces are applied inconsistently, and if you look at [`libbitcoin_*_SOURCES`](../../src/Makefile.am) lists you can see that many libraries pull in files from outside their source directory. But generally when working with libraries, it is good to follow an establishing pattern like:
+
+  - `libbitcoin_node` code lives in `src/node/` in the `node::` namespace
+  - `libbitcoin_wallet` code lives in `src/wallet/` in the `wallet::` namespace
+  - `libbitcoin_ipc` code lives in `src/ipc/` in the `ipc::` namespace
+  - `libbitcoin_util` code lives in `src/util/` in the `util::` namespace
+  - `libbitcoin_consensus` code lives in `src/consensus/` in the `Consensus::` namespace
+
+## Dependencies
+
+- Libraries should generally be careful about what other libraries they depend on, and only reference symbols following the arrows shown in the dependency graph below:
+
+```mermaid
+graph TD;
```

Comment:
This looks a bit scary with all these lines. It's probably better to use (a reasonably chosen set of) fixed positions and straight lines.

Reply:
Yeah this graph is atrocious. I thought mermaid might do a better job with it. I think the markdown source code is more undertandable than the graphic, but I will look for mermaid options (straight lines!) which might make this more readable.

Reply:
Flow chart still looks confusing. I tried creating some flowcharts using [mermaid live editor](https://mermaid-js.github.io/mermaid-live-editor/) based on syntax shared [here](https://mermaid-js.github.io/mermaid/). I was able to simplify bitcoind, bitcoin-qt and bitcoin-wallet but not other things (used different color for binaries):

```md
flowchart TB
    subgraph 1
    d[bitcoind]-->libbitcoin_wallet
    end
    subgraph 2
    qt[bitcoin-qt]-->libbitcoin_node
	qt[bitcoin-qt]-->libbitcoinqt
	qt[bitcoin-qt]-->libbitcoin_wallet
    end
    subgraph 3
    w[bitcoin-wallet]-->libbitcoin_wallet
	w[bitcoin-wallet]-->libbitcoin_wallet_tool
    end

    classDef teal fill:#033E3E,stroke:#AFDCEC,stroke-width:1px;
    class d,qt,w teal
```

```mermaid
flowchart TB
    subgraph 1
    d[bitcoind]-->libbitcoin_wallet
    end
    subgraph 2
    qt[bitcoin-qt]-->libbitcoin_node
	qt[bitcoin-qt]-->libbitcoinqt
	qt[bitcoin-qt]-->libbitcoin_wallet
    end
    subgraph 3
    w[bitcoin-wallet]-->libbitcoin_wallet
	w[bitcoin-wallet]-->libbitcoin_wallet_tool
    end

    classDef teal fill:#033E3E,stroke:#AFDCEC,stroke-width:1px;
    class d,qt,w teal
```

Reply:
> though I avoided black text on teal background, because that was not readable on my PC

Just realized it looks different in github (light theme). Text is white in github (dark theme)

> Probably there is room to do fancier things here in the future, but for now I'm happy if graph just shows what symbol dependencies between libraries are allowed.

Makes sense.